docs/dcp2_system_design.rst

```diff
@@ -0,0 +1,1465 @@
+   provide idempotence (ability to rerun the import without having
+   to clear the TDR tables) and provenance (ability to trace data
+   back from TDR to DSS).
+
+-  There is also a ``version`` PK column, a BQ TIMESTAMP. An
+   update to a subgraph manifests as a new row. The new row has the
+   same value in ``links_id`` as the original row and an
+   increased value in ``version``. The two rows also differ
+   in their ``rowid``. For links imported from the DSS, this
+   column is set to the version of the bundle containing the
+   corresponding ``links.json`` metadata file in the DSS.
+
+TDR snapshots contain the same table, but just like entity tables in
+snapshots, only one version of a subgraph may exist and the
+``links_id`` column is a unique key for all rows in the
+``links`` table of a snapshot.
+
+Subgraph reconstruction
+~~~~~~~~~~~~~~~~~~~~~~~
+
+In order for metadata consumers to be able to reconstruct the HCA
+metadata subgraphs from the contents of a TDR dataset or snapshot, the
+schema for ``links.json`` was updated so that each reference
+from a link to a process, input or output is qualified with the concrete
+type of the entity, enabling metadata consumers to identify the name of
+the table from which to read the respective process, input or output.
+Under the revised schema, an example entry in ``links.json``
+looks as follows::
+
+    {
+        "process_id": "b7a172d6-dbb1-41f3-8ae4-7807e1eca803", # renamed from process!!!
+        "process_type": "analysis_process",
+        "inputs": [
+            {
+                "input_type": "sequence_file",
+                "input_id": "6f725a94-5c81-45e5-8d16-96520aa99703"
+            }
+        ],
+        "outputs": [
+            {
+                "output_type": "analysis_file",
+                "output_id": "2e072336-b906-4c9b-a475-ca03f51c8452"
+            }
+        ],
+        "protocols": [
+            {
+                "protocol_type": "analysis_protocol",
+                "protocol_id": "4d6f7580-ce81-4a81-9c2c-872fcb23b7cd"
+            }
+        ]
+    }
+
+The steps to reconstruct a subgraph from the tables in a TDR *dataset*
+are as follows:
+
+#. Given a ``links_id`` (the UUID of a subgraph) and
+   ``version``, fetch the row with that ``links_id``
+   and ``version`` from the ``links`` table.
+
+#. Read that row's ``content`` column and iterate over the items
+   of the ``links`` property (a list) of the JSON document in
+   ``content``. For each item aka link,
+
+   a. read the ``process`` property, and extract the schema type
+      from the ``process_type`` property and the process
+      UUID from the ``process_id``. Query the TDR table that
+      corresponds to the schema type and fetch all rows where
+      ``{schema_type}_id`` equals the process UUID. Pick the row
+      with the highest version.
+
+   b. read the ``inputs``, ``outputs`` and
+      ``protocols`` properties (they're all lists). For each
+      input, output and protocol,
```

Comment:
This seems to be determined by the rendering engine (Github in this case). I thought the manual numbering would come through, but since it didn't, I changed it all to auto-numbering.